File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        if(num.size()==k)return "0";
        string res;
        for(int i=0;i<num.size();i++)
        {
            if(k&&!res.empty()&&res.back()>num[i])
            {
                res.pop_back();
                i--;
                k--;
            }
            else
            {
                res+=num[i];
            }
        }
        int rem=0;
        for(int i=0;i<res.size();i++)
        {
            if(res[i]!='0')break;
            rem++;
        }
        res=res.substr(rem);
        res=res.substr(0,max(0,(int)res.size()-k));
        return res.empty()?"0":res;
    }
};

File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        Stack<Character> st = new Stack<>();
        for(int i=0;i<num.length();i++)
        {
            while(st.size() > 0 && k > 0 && st.peek() > num.charAt(i))
            {
                st.pop();
                k--;
            }
            st.push(num.charAt(i));
        }
        StringBuilder temp = new StringBuilder();
        while(k>0)
        {
            k--;
            st.pop();
        }
        while(st.size()>0)
        {
            temp.append(st.pop());
        }
        temp = temp.reverse();
        StringBuilder res = new StringBuilder();
        boolean skip = true;
        for(int i=0;i<temp.length();i++)
        {
            if(temp.charAt(i)=='0' && skip)
            {
                continue;
            }
            else
            {
                skip = false;
            }
            res.append(temp.charAt(i));
        }
        if(res.length()==0)
            return "0";
        return res.toString();
    }
}